### ISSUE TITLE
ctx_vqgan decoder produces silently wrong output on Apple Silicon (PyTorch MPS)

### ISSUE THREAD

**Issue by aviadarn** [thuml/iVideoGPT#29]
Thanks for the great codebase — I fine-tuned the action-conditioned model on BridgeData V2 on a MacBook and hit a silent numerical bug worth documenting.

**Environment:** Apple Silicon macOS, Python 3.9, torch 2.2.1, `PYTORCH_ENABLE_MPS_FALLBACK=1`, model `thuml/ivideogpt-oxe-64-act-free`.

**Symptom:** a pure `tokenize()` → `detokenize()` roundtrip through `CompressiveVQModel` returns badly corrupted reconstructions on MPS, with no error or warning:

| Sample | CPU recon PSNR | MPS recon PSNR |
|---|---|---|
| `inference/samples/fractal_sample.npz` | 30.76 | 16.65 |
| BridgeData V2 episode (64×64) | 23.50 | 14.35 |

**Isolation:** the bug is in the decoder only —
- token ids from `tokenize()` are bit-identical between CPU and MPS (0.00% mismatch),
- transformer logits (act-free and `HeadModelWithAction`) match CPU to ~1e-6,
- `detokenize()` of the *same* token ids differs: CPU output is clean, MPS output is ~12–16 PSNR mush.

**Impact:** training on MPS is unaffected (cross-entropy is computed on token ids), but every decoded validation GIF and pixel metric logged during MPS training is corrupted — it looks exactly like the model diverging, which cost me a while.

**Workaround:** keep a CPU copy of the tokenizer and run `detokenize()` on it; generate on MPS as usual. Implemented in my fork's eval script: https://github.com/aviadarn/iVideoGPT/blob/main/inference/eval_bridge.py

Minimal repro:

```python
import torch, numpy as np
from ivideogpt.vq_model import CompressiveVQModel

tok = CompressiveVQModel.from_pretrained('pretrained_models/ivideogpt-oxe-64-act-free',
                                         subfolder='tokenizer', low_cpu_mem_usage=False).eval()
x = torch.rand(1, 16, 3, 64, 64)  # or a real sample; effect is much larger on real frames
with torch.no_grad():
    t_cpu, _ = tok.tokenize(x, 2)
    r_cpu = tok.detokenize(t_cpu, 2).clamp(0, 1)
    tok_mps = tok.to('mps')
    t_mps, _ = tok_mps.tokenize(x.to('mps'), 2)
    r_mps = tok_mps.detokenize(t_mps.to('mps'), 2).clamp(0, 1).cpu()
print('token ids equal:', torch.equal(t_cpu, t_mps.cpu()))       # True
print('pixel max diff :', (r_cpu - r_mps).abs().max().item())    # large
```

I haven't isolated the exact op yet (suspects: decoder attention/GroupNorm/upsample path under MPS). Happy to dig further or PR a docs note / CPU-decode fallback if useful.

**Comment by aviadarn** [comment]
Found the exact op, and it is not the decoder architecture but a PyTorch MPS kernel bug hit through a memory-layout detail.

**Root cause.** A layer-by-layer CPU-vs-MPS comparison of `detokenize()` (all 184 leaf modules) shows the first divergence at the `SiLU` right after the mid-block: inputs identical to 2e-6, outputs differ by 100 %. That input is a **channels-last view** (strides `(HW·C, 1, W·C, C)`) produced by the diffusers mid-block attention's `transpose(-1,-2).reshape(B,C,H,W)` and by `CrossAttentionBlock`'s `permute(0,2,1).reshape(z.shape)`; convolutions keep that layout downstream. On torch <= 2.2 the MPS `SiLU` kernel silently returns wrong values for channels-last inputs: running that single module standalone on MPS gives rel diff 1.0 on the strided tensor and 2e-8 on `.contiguous()` of the same values. The encoder path never creates this layout, which is why token ids match bit for bit.

**Fix.** Three `.contiguous()` calls in `conditional_vae.py`: https://github.com/thuml/iVideoGPT/pull/30. Before: 14.88 dB MPS-vs-CPU on torch 2.2.1; after: 120 dB (max diff 0.0000). PyTorch 2.14 already gives identical output without the patch, so upgrading torch is the other way out; the `scripts/check_mps_decoder.py` in the PR reproduces both states in one command.
